### PR TITLE
Generic `http` datasource

### PR DESCRIPTION
**Pull request by igorsereda**

### CODE REVIEW

**Comment by igorsereda** [comment]
[Generic HTTP datasource [4.3.0]](https://trello.com/c/WEaT1pKQ/449-generic-http-datasource-430)

File: src/dipdup/datasources/const.py [hunk 1]
```diff
@@ -0,0 +1,7 @@
+class DatasourceKind:
```
Comment (droserasprout) [inline src/dipdup/datasources/const.py:1]: `kind` is just a hint for Pydantic to resolve unions in config correctly. It shouldn't be used anywhere except a single `Literal['kind']` typehint in classes that inherit `DatasourceConfig`.

File: src/dipdup/datasources/datasource.py [hunk 2]
```diff
@@ -38,6 +40,26 @@ def set_logger(self, name: str) -> None:
...
+        retry_sleep=1,
+        ratelimit_rate=0,
+        ratelimit_period=0,
+    )
+
+    def __init__(self, http_config: Optional[HTTPConfig] = None) -> None:
+        super().__init__('', self._default_http_config.merge(http_config))
+        self._logger = _logger
+
+    async def get(self, url: str, cache: bool = False, weight: int = 1, **kwargs):
```
Comment (droserasprout) [inline src/dipdup/datasources/datasource.py:56]: Nice idea. I think it's better to add helpers for all basic HTTP methods in `HttpGateway`. One may say it's not necessary to add a bunch of methods that hide a single argument, but this is expected behavior for `requests` and `aiohttp` users.

File: src/dipdup/datasources/factory.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+        datasource.set_logger(name)
+        datasource.set_user_agent(config.package)
+
+        return datasource
+
+    @classmethod
+    def _build_datasource(cls, name: str, config: DipDupConfig) -> Datasource:
+        datasource_config = config.get_datasource(name)
+
+        if datasource_config.kind == DatasourceKind.tzkt:
```
Comment (droserasprout) [inline src/dipdup/datasources/factory.py:35]: As mentioned above, please, don't use this field.

File: src/dipdup/datasources/factory.py [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+from dipdup.datasources.coinbase.datasource import CoinbaseDatasource
+from dipdup.datasources.const import DatasourceKind
+from dipdup.datasources.datasource import Datasource
+from dipdup.datasources.datasource import HttpDatasource
+from dipdup.datasources.ipfs.datasource import IpfsDatasource
+from dipdup.datasources.metadata.datasource import MetadataDatasource
+from dipdup.datasources.tzkt.datasource import TzktDatasource
+
+
+class DatasourceFactory:
```
Comment (droserasprout) [inline src/dipdup/datasources/factory.py:20]: I don't like the idea to add a 105 LOC module for a single factory. Something like this should be more than enough:

```python
@classmethod
def build(cls, name: str, config: DatasourceConfig) -> Datasource:
    datasource = cls._build_datasource(config)
    datasource.set_logger(name)
    datasource.set_user_agent(config.package)

@classmethod
def _build_datasource(config: DatasourceConfig) -> Datasource:
    if isinstance(config, TzktDatasourceConfig):
        return TzktDatasource(
            url=datasource_config.url,
            http_config=datasource_config.http,
            merge_subscriptions=config.advanced.merge_subscriptions,
        )
    if isinstance(config, ...):
        ...
    raise NotImplementedError
```

**Review by droserasprout — CHANGES_REQUESTED** [review]

File: src/dipdup/datasources/datasource.py (hunk 2 above)
Comment (igorsereda) [reply src/dipdup/datasources/datasource.py:56]: The only reason to add this method here is that the request method of `HttpGateway` joins path from the `url` call parameter and domain from the `url` datasource settings, which is not provided for `HttpDatasource` by design.

**Pull request merged**